ement-room.el

```diff
@@ -2817,151 +2809,113 @@ MSG is the message STRING was contained in."
+         (saved-buffer (current-buffer)))
+    (unless (or ement-room--preview-stop
+                (map-elt (ement-event-local related-event)
+                         'url-preview-n))
+      (ement-api ement-session "preview_url"
+        :endpoint-category "media" :params `((url ,link))
+        :then (lambda (data)
+                (when (buffer-live-p saved-buffer)
+                  (with-current-buffer saved-buffer
+                    (ement-room--insert-preview msg related-event num (ement-room--cleanse-url-preview data link)))))
+        :else (lambda (why) (ement-debug "Error in fetching preview: " why))))))
+
+(defun ement-room--cleanse-url-preview (data link)
+  "Ensure a url preview DATA always has a title and associated LINK."
+  (unless (alist-get 'og:url data) (push (cons 'og:url link) data))
+  (unless (alist-get 'og:title data)
+    (push (cons 'og:title (replace-regexp-in-string "\\<https?://\\|/.*$\\|/$" "" link)) data))
+  data)
+
+(defun ement-room--insert-preview (msg related num data)
   "Insert link preview for DATA into MSG from node RELATED if it isn't already present.
-BUF is the room buffer this link is from. NUM is the index for this link."
-  (cond ((alist-get 'errcode data) (message "Ement: You are being rate limited."))
-        (data
-         (with-current-buffer buf
-           (let* ((id (ement-event-id msg))
-                  (related-node
-                   (ement-room--ewoc-node-before ement-ewoc
-                                                 msg
-                                                 (lambda (a b)
-                                                   (string= (ement-event-id a)
-                                                            (ement-event-id b)))
-                                                 :pred #'ement-event-p))
-                  (key (intern (concat "preview-" (number-to-string num)))))
-             (unless (alist-get key (map-elt (ement-event-local related)
-                                             (intern (concat "url-previews-" id))))
-               (cl-pushnew (cons key data) (map-elt (ement-event-local related)
-                                                    (intern (concat "url-previews-" id)))))
-             (unless (> num (or (car-safe (map-elt (map-elt (ement-event-local related)
-                                                            (intern (concat "url-previews-" id)))
-                                                   'url-preview-n))
-                                most-positive-fixnum))
-               (cl-pushnew `(url-preview-n ,num)
-                           (map-elt (ement-event-local related)
-                                    (intern (concat "url-previews-" id)))))
-             (ewoc-invalidate ement-ewoc related-node))))))
+NUM is the index for this link."
+  (cond ((alist-get 'errcode data) (progn
+                                     (message "Ement: You are being rate limited. Waiting 10s before trying again.")
```

Comment:
I appreciate your thoroughness here, but let's leave retrying out of the code for now, to keep it simple.  These requests should fail rarely, and when they do, they aren't important.  I'd rather implement retries in a more general way at a higher level, e.g. in `plz` or `ement-api`.